#!/bin/bash
#SBATCH --job-name=prep_contam_had_crashed_ben_tweaked_script
#SBATCH --mail-user=mb2836@cornell.edu
#SBATCH --mail-type=ALL
#SBATCH -c 90
#SBATCH --mem=1000G
#SBATCH --partition=regular
#SBATCH --qos=regular
#SBATCH -o ../Logs/%x_%j.out
#SBATCH -e ../Logs/%x_%j.err

# Fasta generated by previous step
fasta=

# Fastq of adaptor filtered sequences
fastq=

# Path to ncbi nt database 
nt=

# Path to minimap2
minimap=

# Path to samtools
samtools=

# Path to singualrity container with BUSCO
busco=

# Path to blastn
blast=

# Map the reads back to hifiasm build
# ${minimap} -x map-hifi -a -t 90 ${fasta} ${fastq} | ${samtools} view --threads 90 -b - | ${samtools} sort -o HiFi.to.Initial_Build.bam --threads 90

# ${samtools} flagstat --threads 90 HiFi.to.Initial_Build.bam > HiFi.to.Initial_Build.bam.mapstats

# Make sure that all reads were mapped
# echo "Total number of HiFi reads" > HiFi.mapping_check.txt
# wc -l ${fastq} | awk '{print $1/4}' >> HiFi.mapping_check.txt
# echo "Total primary maps in HiFi.to.Initial_Build.bam" >> HiFi.mapping_check.txt
# head -2 HiFi.to.Initial_Build.bam.mapstats | tail -1 >> HiFi.mapping_check.txt

# Calculate coverage from bam file using samtools
# ${samtools} coverage HiFi.to.Initial_Build.bam > HiFi.to.Initial_Build.bam.coverage

# Run BUSCO on the original assembly
singularity run \
    --bind /local/storage/Projects/p_reticulata_assembly/ \
    --pwd  $(pwd) \
    ${busco} busco \
    -c 90 \
    -i ${fasta} \
    -f \
    -l /local/storage/Projects/p_reticulata_assembly/2_contam_removal/2_prep/actinopterygii_odb10 \
    --out Initial_Build.busco.preclean \
    --mode genome \
    --download_path . \


# Run Megablast of genome to the nt database
# ${blast} -db /local/storage/Databases/NCBI_nt/nt \
#    -query ${fasta} \
#    -outfmt "6 qseqid staxids bitscore std" \
#    -max_target_seqs 10 \
#    -max_hsps 1 \
#    -evalue 1e-25 \
#    -num_threads 90 \
#    -out Initial_Build.ncbi.blastn.out

# Make a yaml file to use with blobtools
echo -e "assembly:\n  level: contig\n  prefix: renome\ntaxon:\n  name: Poecilia_reticulata" > Initial_Build.meta.yaml

